Added try_recv & try_send to UnixDatagram

jean-airoldie opened the pull request:
This allows nonblocking sync send & recv operations on the socket.

### Motivation
This is useful for draining the socket's buffer before dropping it. To do so, one would shutdown the writing end of the peer than `try_recv` until it errors.

carllerche commented:
This seems fine to me. Looks like Ci is failing.

Thoughts @tokio-rs/maintainers ?

jean-airoldie commented:
Seems like freebsd doesn't return WouldBlock for some reason.

jean-airoldie commented:
Turns out that freebsd return `Os { code: 55, kind: Other, message: "No buffer space available" }'` instead of `WouldBlock` when the datagram queue is full.

Ralith commented:
> This is useful for draining the socket's buffer before dropping it.

When is that needed?

kleimkuhler commented on tokio-net/src/uds/datagram.rs:
```diff
@@ -75,6 +75,11 @@ impl UnixDatagram {
 
+    /// Try to send a datagram from the socket without blocking.
+    pub fn try_send(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.io.get_mut().send(buf)
```
```suggestion
        self.io.get_ref().send(buf)
```

Should be okay here!

jean-airoldie commented:
> When is that needed?

I'm passing shared memory objects so I need to properly deallocate them when dropping the channel.

carllerche commented:
I'm good w/ this. Could we fix the merge conflicts and I would :heart: if the API docs could be padded out a bit and have an example.

jean-airoldie commented:
Almost forgot about this PR.

I added the `try_send_to` and `try_recv_from` methods as well. I improved the documentation and added 2 basics tests in the doc to show usage.

kleimkuhler commented:
@jean-airoldie Sorry there has not been any new activity on this. It should still be good to go, but it looks like `rusfmt` is failing. The build was cleared, but it looks like `try_send_to` needs a comma after `P: AsRef<Path>`.

Once the build is passing again it should be 👍

jean-airoldie commented:
I'll rebase against master to see if this fixes it. If I run rustfmt on my machine I see nothing wrong.

jean-airoldie commented:
I manually edited it. Its weird that I doesn't show up on my machine because I'm using the same rustfmt version than the CI.

kleimkuhler commented:
@jean-airoldie When manually checking formatting I usually do `cargo fmt --all -- --check`, which on your branch before the previous commit did *not* show any errors. If that is how you were checking, I don't think an error shows up because the `datagram.rs` file is not included in the default features.

CI seems to address this by passes all the files manually with a `find` command.

Darksonn commented:
This would also make sense to add to the split `UnixDatagram`. However, I don't think it should block merging this PR.
